<a href="https://colab.research.google.com/github/ahdi3/KomputasiIntelegensiaTasks7/blob/main/TaskWeek7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama: Ahdi Nazzalul Ahsan

NPM: 2206812956

Original Source from Hugging Face: https://huggingface.co/ahmedrachid/FinancialBERT-Sentiment-Analysis

# 1. Run the model on google colab and try to predict a sentence

In [1]:
# Install transformers if it's not installed
!pip install transformers torch

# Import necessary libraries
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch

# Load the model and tokenizer from Hugging Face
model_name = "ahmedrachid/FinancialBERT-Sentiment-Analysis"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

# Define a function to analyze the sentiment
def analyze_sentiment(text):
    inputs = tokenizer(text, return_tensors="pt", truncation=True, padding=True, max_length=512)
    with torch.no_grad():
        outputs = model(**inputs)
        logits = outputs.logits
        predicted_class = torch.argmax(logits).item()

    labels = {0: "Negative", 1: "Neutral", 2: "Positive"}
    return labels[predicted_class]

# Test the model with a financial sentence
text = "The company's stock price increased significantly after the earnings report."
sentiment = analyze_sentiment(text)

print(f"Text: {text}")
print(f"Sentiment: {sentiment}")


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/369 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/226k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/464k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


config.json:   0%|          | 0.00/789 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/439M [00:00<?, ?B/s]

Text: The company's stock price increased significantly after the earnings report.
Sentiment: Positive


Here's a code that demonstrates how to load and use the `ahmedrachid/FinancialBERT-Sentiment-Analysis` model from Hugging Face for sentiment analysis in Google Colab:

1.   First, ensure that you have the `transformers` and `torch` libraries installed.
2.   Load the pre-trained FinancialBERT model and tokenizer.
3.   Perform sentiment analysis on a financial sentence.



Steps:
1. Install transformers and torch.
2. Load the model and tokenizer from Hugging Face's model hub.
3. Define a function analyze_sentiment that tokenizes input text, passes it to the model, and then predicts the sentiment.
4. Test the function on a sample financial sentence.

In this example:
*   Sentiment labels are 0 for Negative, 1 for Neutral, and 2 for Positive.


In this example text is: **The company's stock price increased significantly after the earnings report**

And Sentiment's output is **Positive**

In [2]:
from transformers import pipeline

model_name = "ahmedrachid/FinancialBERT-Sentiment-Analysis"
sentiment_analyzer =  pipeline("sentiment-analysis", model=model_name)

sample_text = "The company's stock price increased significantly after the earnings report."
result = analyze_sentiment(sample_text)

print(result)

Positive


/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


In [3]:

# Import necessary libraries
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch

# Load the model and tokenizer from Hugging Face
model_name = "ahmedrachid/FinancialBERT-Sentiment-Analysis"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

# Function to analyze sentiment
def analyze_sentiment(text):
    # Tokenize the input text
    inputs = tokenizer(text, return_tensors="pt", truncation=True, padding=True, max_length=512)

    # Disable gradient calculations (inference mode)
    with torch.no_grad():
        outputs = model(**inputs)
        logits = outputs.logits

        # Get the index of the highest score (sentiment class)
        predicted_class = torch.argmax(logits).item()

    # Define label mapping
    labels = {0: "Negative", 1: "Neutral", 2: "Positive"}
    return labels[predicted_class]

# Function to test the model with new text
def test_model_with_new_text():
    while True:
        # Prompt the user to enter text or type 'exit' to stop
        text = input("Enter financial text for sentiment analysis (or type 'exit' to quit): ")

        # Exit the loop if the user types 'exit'
        if text.lower() == 'exit':
            print("Exiting sentiment analysis.")
            break

        # Analyze the sentiment of the entered text
        sentiment = analyze_sentiment(text)

        # Display the result
        print(f"Sentiment: {sentiment}\n")

# Call the function to test with new text
test_model_with_new_text()


Enter financial text for sentiment analysis (or type 'exit' to quit): The company's profits have significantly decreased this quarter.
Sentiment: Negative

Enter financial text for sentiment analysis (or type 'exit' to quit): The market reacted positively to the new product launch.
Sentiment: Positive

Enter financial text for sentiment analysis (or type 'exit' to quit): There has been no significant change in the stock price this month.
Sentiment: Neutral

Enter financial text for sentiment analysis (or type 'exit' to quit): exit
Exiting sentiment analysis.


How it works:
1.   `The test_model_with_new_text()` function continuously prompts the user to input financial text for sentiment analysis.
2.   The user can input a text, and the `analyze_sentiment` function will process it using the model and return the sentiment (Negative, Neutral, or Positive).
3. If the user types "exit", the loop will stop and the function will terminate.



In this example:
1. The first input "The company's profits have significantly decreased this quarter." results in a **Negative** sentiment.
2. The second input "The market reacted positively to the new product launch." gives a **Positive** sentiment.
3. The third input "There has been no significant change in the stock price this month." results in a **Neutral** sentiment.
4. Typing "exit" ends the interaction.

# 2. Check the model’s accuracy using a synthetic dataset

In [4]:
# Import necessary libraries
import numpy as np
from sklearn.metrics import accuracy_score

# Generate a synthetic dataset
# Example financial sentences with corresponding sentiment labels (0: Negative, 1: Neutral, 2: Positive)
synthetic_data = [
    ("The company's profits have significantly decreased this quarter.", 0),  # Negative
    ("The market reacted positively to the new product launch.", 2),           # Positive
    ("There has been no significant change in the stock price this month.", 1), # Neutral
    ("Investors are worried about the company's future.", 0),                  # Negative
    ("The stock is expected to perform well in the coming months.", 2),       # Positive
    ("Earnings reports showed consistent performance.", 1),                    # Neutral
    ("The unexpected loss led to a drastic drop in stock prices.", 0),        # Negative
    ("Analysts remain cautiously optimistic about growth.", 2),                # Positive
    ("Market trends indicate stability.", 1),                                  # Neutral
    ("The company failed to meet revenue expectations.", 0),                  # Negative
]

# Separate sentences and labels
texts, true_labels = zip(*synthetic_data)

# Make predictions using the model
predicted_labels = [analyze_sentiment(text) for text in texts]

# Map predicted labels back to numeric values
label_mapping = {"Negative": 0, "Neutral": 1, "Positive": 2}
predicted_numeric_labels = [label_mapping[label] for label in predicted_labels]

# Calculate accuracy
accuracy = accuracy_score(true_labels, predicted_numeric_labels)

# Display the results
print("Predicted Labels:", predicted_labels)
print("True Labels:", ["Negative" if label == 0 else "Neutral" if label == 1 else "Positive" for label in true_labels])
print(f"Accuracy of the model on the synthetic dataset: {accuracy * 100:.2f}%")


Predicted Labels: ['Negative', 'Positive', 'Neutral', 'Negative', 'Positive', 'Positive', 'Negative', 'Positive', 'Positive', 'Negative']
True Labels: ['Negative', 'Positive', 'Neutral', 'Negative', 'Positive', 'Neutral', 'Negative', 'Positive', 'Neutral', 'Negative']
Accuracy of the model on the synthetic dataset: 80.00%


Explanation:

1. **Synthetic Dataset Creation**: A list of tuples is created, where each tuple contains a financial sentence and its corresponding label (0 for Negative, 1 for Neutral, 2 for Positive).

2. **Prediction**: The model analyzes each sentence in the synthetic dataset using the `analyze_sentiment` function.

3. **Mapping Labels**: The predicted labels are converted back into numeric values for accuracy calculation.

4. **Accuracy Calculation**: The accuracy of the model is calculated using `sklearn.metrics.accuracy_score` by comparing the true labels and the predicted labels.

5. **Results Display**: The predicted labels, true labels, and accuracy percentage are printed out.

And the result is accuracy of the model on the synthetic dataset: 80.00%

This can be seen based on 10 sentiments, only 8 can be predicted correctly.

# 3. Implement attention transformer on that model

In [5]:
import torch
import torch.nn as nn
from transformers import BertModel, BertTokenizer

# Define a simple transformer model with attention
class SimpleTransformer(nn.Module):
    def __init__(self, bert_model_name):
        super(SimpleTransformer, self).__init__()
        self.bert = BertModel.from_pretrained(bert_model_name)
        self.fc = nn.Linear(self.bert.config.hidden_size, 3)  # 3 classes: Negative, Neutral, Positive

    def forward(self, input_ids, attention_mask, token_type_ids=None):
        outputs = self.bert(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids)
        # Take the output of the [CLS] token for classification
        cls_output = outputs.last_hidden_state[:, 0, :]
        logits = self.fc(cls_output)
        return logits

# Load the tokenizer for the transformer model
bert_model_name = "bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(bert_model_name)

# Function to analyze sentiment using the custom transformer
def analyze_sentiment_transformer(model, text):
    inputs = tokenizer(text, return_tensors="pt", truncation=True, padding=True, max_length=512)
    with torch.no_grad():
        logits = model(**inputs)  # Pass all inputs including token_type_ids
        predicted_class = torch.argmax(logits).item()
    labels = {0: "Negative", 1: "Neutral", 2: "Positive"}
    return labels[predicted_class]

# Initialize the transformer model
transformer_model = SimpleTransformer(bert_model_name)

# Example usage
text = "The company's stock price has dropped significantly this quarter."
sentiment = analyze_sentiment_transformer(transformer_model, text)

print(f"Text: {text}")
print(f"Sentiment: {sentiment}")



tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Text: The company's stock price has dropped significantly this quarter.
Sentiment: Neutral


Explanation:

1.   **SimpleTransformer Class**: This class defines a transformer model that uses a pre-trained BERT model (`bert-base-uncased`) as the backbone. It has a fully connected layer (`fc`) that maps the hidden states of the [CLS] token to the sentiment classes.
2.   **Tokenization**: The `BertTokenizer` is used to tokenize the input text, which prepares it for the model.
3.   **Sentiment Analysis Function**: The `analyze_sentiment_transformer` function takes the transformer model and input text, performs inference, and predicts the sentiment.
4. **Model Initialization**: The `SimpleTransformer` model is initialized using the BERT model.
5. **Example Usage**: You can test the model with a financial sentence, and it will output the predicted sentiment.


In [6]:
import torch
import torch.nn as nn
from transformers import BertModel, BertTokenizer

# Define a simple transformer model with attention
class SimpleTransformer(nn.Module):
    def __init__(self, bert_model_name):
        super(SimpleTransformer, self).__init__()
        self.bert = BertModel.from_pretrained(bert_model_name)
        self.fc = nn.Linear(self.bert.config.hidden_size, 3)  # 3 classes: Negative, Neutral, Positive

    def forward(self, input_ids, attention_mask, token_type_ids=None):
        # If token_type_ids is not provided, we set it to None
        outputs = self.bert(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids)
        # Take the output of the [CLS] token for classification
        cls_output = outputs.last_hidden_state[:, 0, :]
        logits = self.fc(cls_output)
        return logits

# Load the tokenizer for the transformer model
bert_model_name = "bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(bert_model_name)

# Function to analyze sentiment using the custom transformer
def analyze_sentiment_transformer(model, text):
    inputs = tokenizer(text, return_tensors="pt", truncation=True, padding=True, max_length=512)
    with torch.no_grad():
        logits = model(**inputs)  # Pass all inputs including token_type_ids
        predicted_class = torch.argmax(logits).item()
    labels = {0: "Negative", 1: "Neutral", 2: "Positive"}
    return labels[predicted_class]

# Initialize the transformer model
transformer_model = SimpleTransformer(bert_model_name)

# Example usage
examples = [
    "The recent merger has significantly boosted investor confidence.",
    "The quarterly earnings report was disappointing, leading to a drop in stock prices.",
    "Despite the challenges, the company is optimistic about future growth.",
    "There is uncertainty in the market due to economic fluctuations.",
    "The firm's profits are at an all-time high, reflecting strong management.",
]

# Analyze sentiment for each example
for text in examples:
    sentiment = analyze_sentiment_transformer(transformer_model, text)
    print(f"Text: {text}")
    print(f"Sentiment: {sentiment}\n")


Text: The recent merger has significantly boosted investor confidence.
Sentiment: Negative

Text: The quarterly earnings report was disappointing, leading to a drop in stock prices.
Sentiment: Negative

Text: Despite the challenges, the company is optimistic about future growth.
Sentiment: Negative

Text: There is uncertainty in the market due to economic fluctuations.
Sentiment: Positive

Text: The firm's profits are at an all-time high, reflecting strong management.
Sentiment: Negative



Explanation:

1.   **Example Sentences**: We define a list called `examples` that contains various financial statements. Each statement expresses a sentiment related to financial performance, market conditions, or corporate actions.
2.   **LIterating Over Examples**: We use a for `loop` to iterate through each financial statement in the `examples` list.
3. **Sentiment Analysis**: For each statement, we call the `analyze_sentiment_transformer` function. This function tokenizes the text, passes it through the transformer model, and retrieves the predicted sentiment.
4. **Output**: For each text, the code prints the original statement along with its predicted sentiment. The sentiment can be Negative, Neutral, or Positive.


Notes:
*   **Using the Model**: When calling `model(**inputs)` within the `analyze_sentiment_transformer` function, it will correctly unpack and pass all required arguments, including token_type_ids if they are present.
*   **Handling token_type_ids**: Ensure the `token_type_ids` parameter is optional and defaults to None. This allows the model to work correctly when the tokenizer does not provide it.





Explanations of Sentiment:

*   **Positive Sentiment**: Indicates favorable news or expectations, such as "boosted investor confidence" or "profits are at an all-time high." These phrases convey optimism and success.

*   **Negative Sentiment**: Represents unfavorable information, like "disappointing earnings report" or "drop in stock prices," suggesting that the financial health of the company is in decline.

*   **Neutral Sentiment**: Reflects a lack of strong positive or negative connotations, as seen in statements about "uncertainty in the market." This could mean that the situation is ambiguous without a clear direction.






In [8]:
import torch
from sklearn.metrics import accuracy_score

# Create a synthetic dataset
# Each sentence has a corresponding label: 0 (Negative), 1 (Neutral), 2 (Positive)
synthetic_dataset = [
    ("The company reported a loss this quarter.", 0),  # Negative
    ("Market conditions remain stable.", 1),             # Neutral
    ("Investors are optimistic about the new product.", 2),  # Positive
    ("Sales have decreased, causing concern.", 0),       # Negative
    ("The outlook for next quarter is positive.", 2),    # Positive
    ("Profit margins are unchanged.", 1),                 # Neutral
    ("The stock price has plummeted recently.", 0),      # Negative
    ("Customer satisfaction is at an all-time high.", 2), # Positive
]

# Separate the dataset into inputs and labels
texts, true_labels = zip(*synthetic_dataset)

# Function to evaluate model accuracy
def evaluate_model_accuracy(model, texts, true_labels):
    model.eval()  # Set the model to evaluation mode
    predictions = []

    with torch.no_grad():
        for text in texts:
            sentiment = analyze_sentiment_transformer(model, text)
            # Map sentiment to corresponding label
            predicted_label = { "Negative": 0, "Neutral": 1, "Positive": 2 }[sentiment]
            predictions.append(predicted_label)

    accuracy = accuracy_score(true_labels, predictions)
    return accuracy

# Evaluate the model accuracy on the synthetic dataset
accuracy = evaluate_model_accuracy(transformer_model, texts, true_labels)

print(f"Model Accuracy on Synthetic Dataset After Using Attention Transformer: {accuracy * 100:.2f}%")


Model Accuracy on Synthetic Dataset After Using Attention Transformer: 37.50%


Explanation:
1.   **Synthetic Dataset Creation**: We define a list called `synthetic_dataset`, where each tuple consists of a financial statement and its corresponding sentiment label (0 for Negative, 1 for Neutral, and 2 for Positive).
2.   **Separate Texts and Labels**: We use zip to separate the synthetic dataset into two lists: texts for the sentences and `true_labels` for their corresponding sentiment labels.
3. **Evaluate Model Accuracy Function**: The `evaluate_model_accuracy` function takes the model, the list of texts, and the true labels. It performs the following:

  *   Sets the model to evaluation mode using `model.eval()`.
  *   Initializes an empty list `predictions` to store the model's predicted labels.
  *   Iterates through each text, calling `analyze_sentiment_transformer` to get the predicted sentiment and mapping it to the corresponding label.
  *   Calculates the accuracy using `accuracy_score` from `sklearn.metrics`.

4. **Accuracy Evaluation**: Finally, the code evaluates the model's accuracy on the synthetic dataset and prints the result in percentage format.








Result:

The comparison of model accuracy between before and after using the attention transformer reveals a significant drop in performance:

*   Before using attention transformer: **80% accuracy**
*   After using attention transformer: **37.5% accuracy**

Analysis of the Comparison:

1. Accuracy Drop:
  *   A drastic reduction in accuracy (from 80% to 37.5%) suggests that introducing the attention mechanism negatively impacted the model's ability to classify sentiments accurately.
  *   This indicates that the attention mechanism might not be functioning optimally in this context or that it is not well-suited for the current dataset and task.

2. Possible Reasons for the Accuracy Drop:

  *   Improper Implementation, the attention mechanism might not be correctly integrated into the model, leading to suboptimal learning.
  *   Overfitting or Underfitting, attention transformers are powerful  but if not properly trained or if the dataset is too small, they can cause overfitting or underfitting.
  *   Dataset Mismatch, the synthetic dataset may not have been complex enough to benefit from the attention mechanism, which is designed to capture intricate relationships in the data. In simpler tasks or smaller datasets, attention mechanisms can add unnecessary complexity and cause performance to degrade.
  *   Insufficient Training, the model may require more training or fine-tuning when attention mechanisms are introduced, as they introduce additional parameters and complexity.

3. Task Suitability:

  *   Transformers, especially with attention mechanisms, excel in tasks with large and complex datasets (like machine translation or long-text analysis). For simpler tasks or small datasets like the synthetic one, using such a mechanism might not provide a benefit and can lead to a decrease in performance.
  *   The original setup (without the attention transformer) may already be effective enough for this simpler synthetic dataset.

Conclusion:

In this case, the model without the attention transformer performs better, indicating that attention may not be necessary for this specific task with the current dataset.







